dist/testrunner: Capture number of unittests that passed

leandrolanzieri opened the pull request:
### Contribution description
#13642 introduced the `check_unittests` helper function, to verify the result of unit tests on applications. Some tests (e.g. `gnrc_sixlowpan_iphc_w_vrb`, `gnrc_ipv6_ext_frag`) expect the number of passed tests to be captured. This modifies the regex to do that.

### Testing procedure
Run tests which make use of this function, they should pass.


### Issues/PRs references
#13642

aabadie commented on dist/pythonlibs/testrunner/__init__.py:
```diff
@@ -49,7 +49,7 @@ def run(testfunc, timeout=TIMEOUT, echo=True, traceback=False):
 def check_unittests(child, timeout=TIMEOUT, nb_tests=None):
     _tests = r'\d+' if nb_tests is None else int(nb_tests)
-    child.expect(r'OK \({} tests\)'.format(_tests), timeout=timeout)
+    child.expect(r'OK \(({}) tests\)'.format(_tests), timeout=timeout)
```
I would rather use this on the line above: `'(\d+)'`

miri64 replied:
This then makes the match object in `child.match` very inconsistent, depending if `nb_tests` is provided or not.

aabadie replied:
if nb_tests is provided, this is not a regexp anymore.

miri64 replied:
Of course it is. It just matches to the number provided.

aabadie replied:
if you set nb_tests to a value, but the actual number of successful tests is different, the expect will fail anyway. And it is successful, you already know that number (since it was given as parameter), so there's no need to guess it with a match.

miri64 replied:
Yes, but changing behavior by input is usually a bad idea.

miri64 replied:
This is just about behavior for me and I prefer it to be consistent if possible. There is nothing that speaks against providing this consistency.

aabadie replied:
For me, the `if` above does. When nb_tests is set, you expect an exact string match, there's no regexp involved. IMHO, this is where the consistency is.

aabadie replied:
> changing behavior by input is usually a bad idea.

isn't it the goal of parameters instead ? (e.g. changing parameters ends up in a different behavior)

miri64 replied:
I'm sorry, but your presumption is already wrong. There *is* a regular expression still here that is matched against stdout. It just is not `\d+` (at least one arbitrary digit), but a string representing a specific number. Just because you don't use generic search patterns it does not make it a *not* a regular expression.

```
In [1]: import re                                                                                                                                             

In [2]: re.match("(12345)", "12345")                                                                                                                          
Out[2]: <re.Match object; span=(0, 5), match='12345'>

In [3]: Out[2].group(1)                                                                                                                                       
Out[3]: '12345'

In [4]: re.match("(\d+)", "123")                                                                                                                              
Out[4]: <re.Match object; span=(0, 3), match='123'>

In [5]: Out[4].group(1)                                                                                                                                       
Out[5]: '123'
```

aabadie replied:
I still would like https://github.com/RIOT-OS/RIOT/pull/13665#discussion_r395585375

aabadie replied:
And also `child.expect_exact` if nb_tests is set. So no regexp in this case, just an exact string comparison.

miri64 replied:
Can you please answer me one question: Why do you think your approach is a better approach than having the expect object / side-effects behave consistently on this function call regardless of input?

leandrolanzieri replied:
I don't get what's the harm of capturing the number even if it is known (in the case it is provided).

miri64 replied:
> I don't get what's the harm in capturing the number even if it is known (in the case it is provided).

I guess that is the same question, I was asking ;-).

leandrolanzieri replied:
ping @aabadie ?

aabadie replied:
Still on my position, sorry.

> Why do you think your approach is a better approach than having the expect object / side-effects behave consistently on this function call regardless of input?

The problem is not about treating the same way all kind of inputs but check the input to properly handle it (None => regexp, nb_tests => plain string check). If nb_tests is set, you can use the `expect_exact` match (which I should have done in the first place and which is faster), if not, use `expect` with a regexp.

miri64 replied:
I'm fine with you having a position on that. I just don't understand why you want the behavior like that. IMHO it just makes this function harder to understand and use.

aabadie replied:
>  it just makes this function harder to understand and use.

Why ? If it's hard to understand, add comments.
If you want to return the number of successful tests, if nb_tests is None, just return your proposition below (int(child.match(1)), in call cases, return nb_tests.

miri64 replied:
Okay, but why add a comment if we can have the same intuitively like this? IMHO we are just bike-shedding. `expect_exact` should only have minor speed advantages here and the other "advantage" that one can just carelessly copy-paste is not given here, as this function abstracts that away.

miri64 commented on dist/pythonlibs/testrunner/__init__.py:
```diff
@@ -49,7 +49,7 @@ def run(testfunc, timeout=TIMEOUT, echo=True, traceback=False):
 def check_unittests(child, timeout=TIMEOUT, nb_tests=None):
     _tests = r'\d+' if nb_tests is None else int(nb_tests)
-    child.expect(r'OK \({} tests\)'.format(_tests), timeout=timeout)
+    child.expect(r'OK \(({}) tests\)'.format(_tests), timeout=timeout)
```
```suggestion
    child.expect(r'OK \(({}) tests\)'.format(_tests), timeout=timeout)
    return int(child.match(1))
```
?

leandrolanzieri commented:
I modified the function to always return the amount of passed tests. Also changed the tests that use that function, to check for the returned value instead of getting the captured group.

miri64 commented:
Please squash

aabadie commented on dist/pythonlibs/testrunner/__init__.py:
```diff
@@ -48,8 +48,18 @@ def run(testfunc, timeout=TIMEOUT, echo=True, traceback=False):
+        return int(child.match.group(1))
+    else:
+        _tests = int(nb_tests)
+        child.expect_exact(r'OK \({} tests\)'.format(_tests), timeout=timeout)
```
```suggestion
        child.expect_exact(r'OK \({} tests\)'.format(_tests), timeout=timeout)
```
```suggestion
        child.expect_exact('OK \({} tests\)'.format(_tests), timeout=timeout)
```

leandrolanzieri replied:
Fixed

aabadie commented on dist/pythonlibs/testrunner/__init__.py:
```diff
@@ -48,8 +48,18 @@ def run(testfunc, timeout=TIMEOUT, echo=True, traceback=False):
+    if nb_tests is None:
+        child.expect(r'OK \((\d+) tests\)', timeout=timeout)
+        return int(child.match.group(1))
+    else:
```
The `else` is not needed.

leandrolanzieri replied:
Removed

leandrolanzieri commented:
I forgot to remove the slashes now that the `expect_exact` is a string :sweat_smile:

leandrolanzieri commented:
I fixed that and squashed directly.

leandrolanzieri commented:
Ok now Murdock is happy
